pixeltable/func/callable_function.py

```diff
@@ -0,0 +1,117 @@
+
+    def help_str(self) -> str:
+        res = super().help_str()
+        res += '\n\n' + inspect.getdoc(self.py_fn)
+        return res
+
+    def _as_dict(self) -> Dict:
+        if self.self_path is None:
+            # this is not a module function
+            from .function_registry import FunctionRegistry
+            #id = FunctionRegistry.get().create_function(self.py_fn, self.self_name, self.signature)
+            #return {'id': id.hex}
+            return {}
+        return super()._as_dict()
+
+    @classmethod
+    def _from_dict(cls, d: Dict) -> Function:
+        if 'id' in d:
+            assert 'fqn' not in d
+            from .function_registry import FunctionRegistry
+            return FunctionRegistry.get().get_function(id=UUID(hex=d['id']))
+        return super()._from_dict(d)
+
+
+def make_module_function(
+        py_fn: Callable, *, return_type: ts.ColumnType, param_types: List[ts.ColumnType],
+        module_path: Optional[str] = None, function_name: Optional[str] = None) -> CallableFunction:
+    """Create a CallableFunction that uses the default serialization mechanism (module path).
+
+    Args:
+        module_path: path to the module containing the CallableFunction; if None, uses the module name of the caller
+        function_name: name of the CallableFunction in the module; if None, uses the name of the eval_fn
+    """
+    if module_path is None:
+        module_path = get_caller_module_path()
+    assert module_path is not '__main__'
+
+    if function_name is None:
+        function_name = py_fn.__name__
+    assert function_name != '<lambda>'
+
+    sig = Signature.create(py_fn, False, param_types, return_type)
+    result = CallableFunction(sig, py_fn=py_fn, module_path=module_path, self_name=function_name)
+    FunctionRegistry.get().register_function(f'{module_path}.{function_name}', result)
+    return result
+
+def make_stored_function(
+        py_fn: Callable, *, return_type: ts.ColumnType, param_types: List[ts.ColumnType]) -> CallableFunction:
+    """Create a CallableFunction that serializes to the store"""
+    sig = Signature.create(py_fn, False, param_types, return_type)
+    # we preserve the name
+    return CallableFunction(sig, py_fn=py_fn, self_name=py_fn.__name__)
+
+
+def make_callable_function(
+        py_fn: Callable, *, return_type: ts.ColumnType, param_types: List[ts.ColumnType]) -> CallableFunction:
+    # determine whether this is a named function in a module
+    if py_fn.__module__ == '__main__' or not py_fn.__name__.isidentifier():
+        return make_stored_function(py_fn, return_type=return_type, param_types=param_types)
+    else:
+        return make_module_function(
+            py_fn, return_type=return_type, param_types=param_types, module_path=py_fn.__module__)
+
+
+def udf(*, return_type: ts.ColumnType, param_types: List[ts.ColumnType], py_fn: Optional[Callable] = None) -> Callable:
+    """Returns decorator to create a CallableFunction from a function definition.
```

Comment:
I generally find it cleaner to implement decorators as classes with a `__call__` method, rather than as functions. (Note a decorator can be any `Callable`.)

I'd also prefer to call this `pxt.func` rather than `func.udf`. The term "UDF" sounds archaic to my ears; in my view, they're all just functions.